### PR TITLE
WIP: Add PodPreset support to API

### LINKED ISSUE #266: Add support for Pod-Preset - aka Pod Injection Policy
Just a spot for pointers to the real work so its easy for people to find the other issues, PRs and design docs.

k8s/features issue:  https://github.com/kubernetes/features/issues/162
k8s/community PR: https://github.com/kubernetes/community/pull/254
k8s/kubernetes PR: https://github.com/kubernetes/kubernetes/pull/41931

### PR DESCRIPTION
**Pull request by pmorie**
WIP.

Resolves #266

### CODE REVIEW

File: pkg/apis/servicecatalog/v1alpha1/types.go [hunk 1]
```diff
@@ -274,6 +275,20 @@ type BindingSpec struct {
 	// Names of subordinate objects to create
 	SecretName string `json:"secretName"`
 
+	// TODO: replace with PodPresetTemplateSpec
+
+	// The name of a PodPreset to create.  Must be specified with
```
Comment (luksa) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: Shouldn't this be "Must not be specified with"?

Comment (luksa) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: Ah, nevermind.

Comment (pmorie) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: @luksa 

No - they have to be specified together.  The spec only tells you what should be in the `PodPreset`, you still have to give it a name.  We will have a `PodPresetTemplateSpec` type that combines the spec with an `ObjectMeta` so you can control annotations and labels on the created `PodPreset` - see kubernetes/kubernetes#44791.

File: pkg/apis/servicecatalog/validation/binding.go [hunk 2]
```diff
@@ -52,6 +52,20 @@ func validateBindingSpec(spec *sc.BindingSpec, fldPath *field.Path, create bool)
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("secretName"), spec.SecretName, msg))
 	}
 
+	if spec.AlphaPodPresetName != nil {
+		for _, msg := range apivalidation.NameIsDNSSubdomain(*spec.AlphaPodPresetName, false /* prefix */) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("alphaPodPresetName"), spec.AlphaPodPresetName, msg))
+		}
+
+		if spec.AlphaPodPresetSpec == nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("alphaPodPresentName"), "alphaPodPresetName may not be specified with alphaPodPresetSpec"))
```
Comment (luksa) [inline pkg/apis/servicecatalog/validation/binding.go]: Maybe my brain is fried, but shouldn't this be "alphaPodPresetSpec must be specified with alphaPodPresetName"?

File: pkg/apis/servicecatalog/v1alpha1/types.go (hunk 1 above)
Comment (luksa) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: Yeah, the error message below led me to believe only one needs to be set. But it looks like the error message itself is incorrect. See below.

File: pkg/apis/servicecatalog/validation/binding.go (hunk 2 above)
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/binding.go]: Yeha, it was late, my brain was fried.  Yours is fine.

**Comment by MHBauer** [comment]
If our API depends on an Alpha API does that make our API Alpha? Hmn... :thinking:

File: pkg/apis/servicecatalog/validation/binding.go [hunk 3]
```diff
@@ -52,6 +52,20 @@ func validateBindingSpec(spec *sc.BindingSpec, fldPath *field.Path, create bool)
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("secretName"), spec.SecretName, msg))
 	}
 
+	if spec.AlphaPodPresetName != nil {
```
Comment (MHBauer) [inline pkg/apis/servicecatalog/validation/binding.go]: Should this section be pulled upstream as "validatePodPreset"?

File: pkg/apis/servicecatalog/validation/binding_test.go [hunk 4]
```diff
@@ -142,6 +147,47 @@ func TestValidateBinding(t *testing.T) {
...
+				b.Spec.AlphaPodPresetSpec = &settings.PodPresetSpec{}
+				return b
+			}(),
+			valid: false,
+		},
+		{
+			name: "invalid podPresetname",
+			binding: func() *servicecatalog.Binding {
+				b := validBinding()
+				n := "★"
```
Comment (MHBauer) [inline pkg/apis/servicecatalog/validation/binding_test.go]: ooh, a pretty star ★ :star:

**Comment by pmorie** [comment]
@MHBauer 

>  If our API depends on an Alpha API does that make our API Alpha? Hmn... 🤔

Very good question.  The answer is complex, but basically 'no'.  We have to be able to provide alpha-level fields on APIs that are more mature than alpha, see https://github.com/kubernetes-incubator/service-catalog/issues/753 for more info.

File: pkg/apis/servicecatalog/v1alpha1/types.go [hunk 5]
```diff
@@ -274,6 +275,20 @@ type BindingSpec struct {
...
+	// AlphaPodPresentSpec.  Alpha features may contain bugs and will not be
+	// forward-migratable.
+	// Immutable.
+	AlphaPodPresetName *string `json:"alphaPodPresetName,omitempty"`
+
+	// The spec of the PodPreset to create.  Must be specified with
```
Comment (pwittrock) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: nit: elsewhere comments read 

> X is the Y

instead of

> The Y

File: pkg/apis/servicecatalog/v1alpha1/zz_generated.conversion.go [hunk 6]
```diff
@@ -25,6 +25,7 @@ import (
 	conversion "k8s.io/apimachinery/pkg/conversion"
 	runtime "k8s.io/apimachinery/pkg/runtime"
 	v1 "k8s.io/client-go/pkg/api/v1"
+	settings_v1alpha1 "k8s.io/client-go/pkg/apis/settings/v1alpha1"
```
Comment (pwittrock) [inline pkg/apis/servicecatalog/v1alpha1/zz_generated.conversion.go]: Does it make sense to split generated stuff into a separate commit?

File: pkg/apis/servicecatalog/types.go [hunk 7]
```diff
@@ -272,6 +273,20 @@ type BindingSpec struct {
 	// Names of subordinate objects to create
 	SecretName string
 
+	// TODO: replace with PodPresetTemplateSpec
+
+	// The name of a PodPreset to create.  Must be specified with
+	// AlphaPodPresentSpec.  Alpha features may contain bugs and will not be
+	// forward-migratable.
+	// Immutable.
+	AlphaPodPresetName *string
```
Comment (pwittrock) [inline pkg/apis/servicecatalog/types.go]: Should these be encapsulated in a single struct that contains both?

File: pkg/apis/servicecatalog/validation/binding.go [hunk 8]
```diff
@@ -52,6 +52,20 @@ func validateBindingSpec(spec *sc.BindingSpec, fldPath *field.Path, create bool)
...
+		for _, msg := range apivalidation.NameIsDNSSubdomain(*spec.AlphaPodPresetName, false /* prefix */) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("alphaPodPresetName"), spec.AlphaPodPresetName, msg))
+		}
+
+		if spec.AlphaPodPresetSpec == nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("alphaPodPresentName"), "alphaPodPresetName may not be specified with alphaPodPresetSpec"))
+		}
```
Comment (pwittrock) [inline pkg/apis/servicecatalog/validation/binding.go]: Do we have anything to validate the contents of the spec?

File: pkg/apis/servicecatalog/validation/binding.go [hunk 9]
```diff
@@ -52,6 +52,20 @@ func validateBindingSpec(spec *sc.BindingSpec, fldPath *field.Path, create bool)
...
+	}
+
+	if spec.AlphaPodPresetSpec != nil && spec.AlphaPodPresetName == nil {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("alphaPodPresetSpec"), "alphaPodPresetSpec may not be specified with alphaPodPresetName"))
```
Comment (pwittrock) [inline pkg/apis/servicecatalog/validation/binding.go]: Same issue here:

> alphaPodPresetSpec may not be specified with alphaPodPresetName

to 

> alphaPodPresetSpec may not be specified without alphaPodPresetName

File: pkg/controller/controller.go [hunk 10]
```diff
@@ -1361,6 +1361,30 @@ func (c *controller) injectBinding(binding *v1alpha1.Binding, credentials *broke
 		_, err = c.kubeClient.Core().Secrets(binding.Namespace).Create(secret)
 	}
 
+	if err != nil {
+		return err
+	}
+
+	// PodPresets are immutable, so if a PodPreset already exists, it should
```
Comment (pwittrock) [inline pkg/controller/controller.go]: Does this comment go here or where it is created?

File: pkg/controller/controller.go [hunk 11]
```diff
@@ -1361,6 +1361,30 @@ func (c *controller) injectBinding(binding *v1alpha1.Binding, credentials *broke
...
+	// be an error.
+	if binding.Spec.AlphaPodPresetName == nil {
+		return nil
+	}
+
+	_, err := c.kubeClient.SettingsV1alpha1().PodPresets(binding.Namespace).Get(binding.Spec.AlphaPodPresetName)
+	if err == nil {
+		return fmt.Errorf(`PodPreset "%v/%v" already exists`, binding.Namespace, binding.Spec.AlphaPodPresetName)
```
Comment (pwittrock) [inline pkg/controller/controller.go]: I could be missing context, so this comment may not apply.

This could be a little bit fragile if we introduce any write logic after this that could fail.

If the write logic after this ever failed, it would never be retried because we would also bail here.

Also, does this ever get retried for any other reason.

Maybe we should we may this idempotent, and only throw an error if there is a PodPreset and it is not equal to the one in the spec.

File: pkg/apis/servicecatalog/v1alpha1/types.go [hunk 12]
```diff
@@ -274,6 +275,20 @@ type BindingSpec struct {
 	// Names of subordinate objects to create
 	SecretName string `json:"secretName"`
 
+	// TODO: replace with PodPresetTemplateSpec
+
+	// The name of a PodPreset to create.  Must be specified with
+	// AlphaPodPresentSpec.  Alpha features may contain bugs and will not be
+	// forward-migratable.
+	// Immutable.
+	AlphaPodPresetName *string `json:"alphaPodPresetName,omitempty"`
```
Comment (pwittrock) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: Is name or ObjectMeta better?  Would we ever want to supply labels or annotations as well?

File: pkg/apis/servicecatalog/validation/binding.go (hunk 3 above)
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/binding.go]: Unfortunately we do not export validations at all in client-go.  I opened https://github.com/kubernetes/kubernetes/issues/44995 to try to figure something out about this.

**Comment by pmorie** [comment]
TLDR: I am going to close this PR because I need to rethink the approach due to what we currently can and cannot validate about API objects in the core API server.

Commented at length here: https://github.com/kubernetes-incubator/service-catalog/issues/266#issuecomment-300637343
